core.ts

```diff
@@ -980,6 +988,56 @@ abstract class PinejsClientCoreTemplate<
 		return this.request(params, { method: 'DELETE' });
 	}
 
+	upsert(params: PinejsClientCoreFactory.UpsertParams) {
+		const { id, body, ...restParams } = params;
+
+		if (!isObject(id)) {
+			throw new Error('The id property must be an object');
+		}
+
+		const naturalKeyProps = Object.keys(params.id);
+		if (naturalKeyProps.length === 0) {
+			throw new Error(
+				'The id property must be an object with the natural key of the model',
+			);
+		}
+
+		if ('options' in params) {
+			throw new Error(`'options' are not supported in upserts`);
+		}
+
+		if (body == null) {
+			throw new Error('The body property is missing');
+		}
+
+		const postParams = {
+			...restParams,
+			body: {
+				...body,
+				...id,
+			},
+		};
+		return this.post(postParams).catch(err => {
+			const { body: errorBody } = err;
+			const isUniqueKeyViolationResponse =
+				errorBody == 'Unique key constraint violated' ||
```

Comment:
Is it not possible to check this the status code?

Reply:
I don't think so, since that's how we ended up doing it in the API as well. If there is indeed a better way plz let me know.
See: https://github.com/balena-io/resin-api/pull/986

Reply:
Can you add a 409 check to go with this, and maybe only check the body for `/unique/i`

Reply:
Yea sure. Had in mind to do this on my next hack Friday :+1:

Reply:
Done @Page- . Also squashed and added the required pinejs version in the commit, so that it's mentioned clearly in the CHANGELOG.